### PR TITLE
Documentation: better documentation menu, add guide to writing a client

### PR DESCRIPTION
**Pull request by aarzilli**
```
Documentation: better documentation menu, add guide to writing a client

* Adds some links to the more relevant documentation pages on the main
README.md
* Adds a document informally describing how to write a client of delve
* Adds link to slides describing the architecture of delve

```

### CODE REVIEW

**Comment by aarzilli** [comment]
I am concerned about the quality of some of delve's clients. All of this was already documented in godocs throughout the code, but I want it to collect it into a single document to make it easier to find.

Hopefully this will help with improving the quality of delve's client UIs.

File: Documentation/api/ClientHowto.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+### RPCServer.Command and stale executable files
+
+It's possible (albeit unfortunate) that your user will decide to change the
+source of the program being executed in the debugger, while the debugger is
+running. Because of this it would be advisable that your client check that
+the executable is not stale every time `Command` returns and notify the user
+that the executable being run is stale and line numbers may nor align
+properly anymore.
+
+You can do this bookkeeping yourself, but delve can also help you with the
```
Comment (derekparker) [inline Documentation/api/ClientHowto.md]: nit: capitalization of Delve in this paragraph.

File: Documentation/api/ClientHowto.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+/Users/you/go/src/something/something.go
+```
+
+But `/Users/you/go/src/something` is a symbolic link to
+`/Users/you/projects/golang/something` the path *must* be specified as
+`/Users/you/go/src/something/something.go` and
+`/Users/you/projects/golang/something/something.go` will not be recognized
+as valid.
+
+If you want to let your users specify a breakpoint on a function selecte
```
Comment (derekparker) [inline Documentation/api/ClientHowto.md]: s/selecte/selected/

File: Documentation/api/ClientHowto.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+If you want to support the [same language as dlv's break and trace commands](//github.com/derekparker/delve/tree/master/Documentation/cli/locspec.md)
+ you should call RPCServer.FindLocation and
+then use the returned slice of Location objects to create Breakpoints to
+pass to CreateBreakpoint: just fill each Breakpoint.Addr with the
+contents of the corresponding Location.PC.
+
+## Looking into variables
+
+There are several API entry points to evaluate variables in delve:
```
Comment (derekparker) [inline Documentation/api/ClientHowto.md]: nit: Delve.

File: Documentation/api/ClientHowto.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+```
+
+To load more elements from a map:
+
+```
+fmt.Sprintf("(*(*%q)(%#x))[%d:]", v.Type, v.Addr, len(v.Children)/2)
+```
+
+All the evaluation API calls except ListPackageVars also take a EvalScope
+argument, this specifies which stack frame you are interested into. If you
```
Comment (derekparker) [inline Documentation/api/ClientHowto.md]: s/into/in/

File: Documentation/api/ClientHowto.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+are interested in the topmost stack frame of the current goroutine (or
+thread) use: `EvalScope{ GoroutineID: -1, Frame: 0 }`.
+
+More informations on the expression language interpreted by RPCServer.Eval
```
Comment (derekparker) [inline Documentation/api/ClientHowto.md]: s/informations/information/

File: Documentation/api/ClientHowto.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+```
+$ dlv debug --headless debug
+```
+
+The rest of the command line remains unchanged. You can use `debug`, `exec`,
+`test`, etc... along with `--headless` and they will work. If this project
+is part of a larger IDE integration then you probably have your own build
+system and do not wish to offload this task to delve, in that case it's
+perfectly fine to always use the `dlv exec` command but do remember that:
+1. delve will work much worse unless the binary is compiled specifying `-gcflags='all=-N -l'`
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply Documentation/api/ClientHowto.md]: You're right, I started writing this a long time ago, when it was worse...

**Comment by aarzilli** [comment]
Done. I've also made a [set of example programs](https://github.com/aarzilli/delve_client_testing) to be used to test clients and linked it in the document.

**Comment by aarzilli** [comment]
cc @zolotov  @visualfc @ramya-rao-a @mattn @Carpetsmoker this could be useful to you, particularly the test cases linked at the end.

**Comment by ramya-rao-a** [comment]
@aarzilli Thanks for the notes. Can we add a few more details around gracefully ending the debug process? In VS Code, I have issues mainly around
- clearing the debug binary file that gets created during every debugging session
- ensuring any processes started by the program being debugged (for eg: web servers). For this at the moment I am manually killing the process for the headless server and all its children.

**Pull request merged**